### PR TITLE
pkg/ansible: pass along requests for virtual resources to cluster

### PR DESCRIPTION
**Pull request by shawn-hurley**
**Description of the change:**
* Defining virtual resources as any resource that does not have the
  ability to "get", "list", and "watch" as all are needed for the cache.
* creating a new struct to handle the caching of the cluster APIResources
**Motivation for the change:**

Using a virtual resource in the cache will cause issues because the informer is unable to be started.

### CODE REVIEW

File: pkg/ansible/proxy/cache_response.go [hunk 1]
```diff
@@ -61,25 +62,33 @@ func (c *cacheResponseHandler) ServeHTTP(w http.ResponseWriter, req *http.Reques
...
-			log.Info("Cache miss, can not find in rest mapper", "GVR", gvr)
+			log.Info("Cache miss, can not find in rest mapper")
+			break
+		}
+
+		// Determine if the resource is virtual. If it does not contain all of these then we should not attempt the cache.
+		isVR, err := c.apiResources.IsVirtualResource(k)
+		if err != nil {
+			// break here in case we can not understand if virtual resource or not
+			log.Info("Unable to determine if virual resource", "gvk", k)
```
Comment (jmrodri) [inline pkg/ansible/proxy/cache_response.go]: typo: virual -> virtual

File: pkg/ansible/proxy/cache_response.go [hunk 2]
```diff
@@ -61,25 +62,33 @@ func (c *cacheResponseHandler) ServeHTTP(w http.ResponseWriter, req *http.Reques
...
-		k, err := c.restMapper.KindFor(gvr)
+		k, err := getGVKFromRequestInfo(r, c.restMapper)
 		if err != nil {
 			// break here in case resource doesn't exist in cache
-			log.Info("Cache miss, can not find in rest mapper", "GVR", gvr)
+			log.Info("Cache miss, can not find in rest mapper")
+			break
+		}
+
+		// Determine if the resource is virtual. If it does not contain all of these then we should not attempt the cache.
```
Comment (jmrodri) [inline pkg/ansible/proxy/cache_response.go]: typo? maybe. I think it should say "should not attempt to cache."

File: pkg/ansible/proxy/inject_owner.go [hunk 3]
```diff
@@ -61,6 +62,34 @@ func (i *injectOwnerReferenceHandler) ServeHTTP(w http.ResponseWriter, req *http
...
+		}
+
+		k, err := getGVKFromRequestInfo(r, i.restMapper)
+		if err != nil {
+			// break here in case resource doesn't exist in cache
+			log.Info("Cache miss, can not find in rest mapper")
+			break
+		}
+
+		// Determine if the resource is virtual. If it does not contain all of these then we should not attempt the cache.
```
Comment (jmrodri) [inline pkg/ansible/proxy/inject_owner.go]: Since you're in there, "attempt to cache"

File: pkg/ansible/proxy/proxy.go [hunk 4]
```diff
@@ -261,3 +279,82 @@ func getRequestOwnerRef(req *http.Request) (kubeconfig.NamespacedOwnerReference,
...
+		if err != nil {
+			return false, err
+		}
+		// retry to get the resource
+		a.mu.RLock()
+		apiResource, ok = a.gvkToAPIResource[gvk.String()]
+		a.mu.RUnlock()
+		if !ok {
+			return false, fmt.Errorf("unable to get api resource for gvk: %v", gvk)
+		}
```
Comment (jmrodri) [inline pkg/ansible/proxy/proxy.go]: :+1: This makes sense.

**Comment by fabianvf** [comment]
lgtm, should we add a test case though?

**Pull request merged**